src/watchdog.js

```diff
@@ -0,0 +1,393 @@
+			.then( editor => {
+				this._editor = editor;
+
+				window.addEventListener( 'error', this._boundErrorHandler );
+				this.listenTo( editor.model.document, 'change:data', this._throttledSave );
+
+				this._lastDocumentVersion = editor.model.document.version;
+				this._data = editor.getData();
+
+				return this;
+			} );
+	}
+
+	/**
+	 * Destroys the current editor using the destructor passed to {@link #setDestructor} method.
+	 *
+	 * @returns {Promise.<module:watchdog/watchdog~Watchdog>}
+	 */
+	destroy() {
+		window.removeEventListener( 'error', this._boundErrorHandler );
+		this.stopListening( this._editor.model.document, 'change:data', this._throttledSave );
+
+		return Promise.resolve()
+			.then( () => this._destructor( this._editor ) )
+			.then( () => {
+				this._editor = null;
+
+				return this;
+			} );
+	}
+
+	/**
+	 * Saves the editor data, so it can be restored after the crash even if the data cannot be fetched at
+	 * the moment of a crash.
+	 *
+	 * @private
+	 */
+	_save() {
+		const version = this._editor.model.document.version;
+
+		// Change may not produce an operation, so the document's version
+		// can be the same after that change.
+		if ( version < this._lastDocumentVersion ) {
+			this._throttledSave.cancel();
+
+			return;
+		}
+
+		try {
+			this._data = this._editor.getData();
+			this._lastDocumentVersion = version;
+		} catch ( err ) {
+			console.error(
+				err,
+				'An error happened during restoring editor data. ' +
+				'Editor will be restored from the previously saved data.'
+			);
+		}
+	}
+
+	/**
+	 * Checks if the event error comes from the editor that is handled by the watchdog (by checking the error context) and
+	 * restarts the editor.
+	 *
+	 * @private
+	 * @fires error
+	 * @param {Event} event Error event.
+	 */
+	_handleGlobalErrorEvent( event ) {
+		if ( !event.error.is || !event.error.is( 'CKEditorError' ) ) {
+			return;
+		}
+
+		if ( event.error.context === undefined ) {
+			console.error( 'The error is missing its context and Watchdog cannot restart the proper editor.' );
+
+			return;
+		}
+
+		// In some cases the editor should not be restarted - e.g. in case of the editor initialization.
+		if ( event.error.context === null ) {
+			return;
+		}
+
+		if ( this._isErrorComingFromThisEditor( event.error ) ) {
+			this.crashes.push( {
+				message: event.error.message,
+				source: event.source,
+				lineno: event.lineno,
+				colno: event.colno
+			} );
+
+			this.fire( 'error' );
+
+			if ( this.crashes.length <= this._crashNumberLimit ) {
+				this.restart();
+			}
+		}
+	}
+
+	/**
+	 * Traverses both structures to find out whether the error context is connected
+	 * with the current editor.
+	 *
+	 * @private
+	 * @param {module:utils/ckeditorerror~CKEditorError} error
+	 */
+	_isErrorComingFromThisEditor( error ) {
+		return areConnectedThroughProperties( this._editor, error.context );
+	}
+
+	/**
+	 * A shortcut method for creating the Watchdog.
```

Comment:
If this is the shortcut (shorthand), then what's the full way?

Reply:
Asking, because this wasn't clear to me the first time I saw it.

Reply:
I'll link it here.